### DISCUSSION TITLE
locate service with exclude polygons

### DISCUSSION THREAD

**Discussion by OliverHering** [valhalla/valhalla#4070] [Q&A]
Hi,
I'm trying to get the geometry of the nearest ways to a given point. And the locate service is exactly what I was looking for. 
But if I set a polygon as excluded I would assume, that the response does not contain any ways, which are within this polygon. 
But that's not the case. The locate service seems to ignore the excluded polygon.
Am I missing something?

Here my test request:

`{
    "locations": [
        {
            "lat": 52.354487,
            "lon": 9.730855,
            "radius": 100
        }
    ],
    "costing": "bicycle",
    "exclude_polygons": [
        [
            [
                9.726981,
                52.3613
            ],
            [
                9.729886,
                52.362592
            ],
            [
                9.729927,
                52.362617
            ],
            [
                9.73777,
                52.361528
            ],
            [
                9.740011,
                52.356256
            ],
            [
                9.738185,
                52.352784
            ],
            [
                9.729512,
                52.352581
            ],
            [
                9.726981,
                52.3613
            ]
        ]
    ],
    "units": "km",
    "id": "123",
    "verbose": true
}`

Thank you for your help.

**Comment by nilsnolde** [comment]
No, you're not missing much:) `/locate` is not respecting those exclusions. I see how you'd like those respected with a `radius` set on the locations.

`/locate` is basically only doing the first 1/3 of a `/route` request: it only correlates your locations to the graph, finds all relevant graph objects, then serializes and returns them. When routing, we have to look at the percentage of the correlated road to be able to tell if we'd like to exclude it or not, and that heavily depends whether you're doing a forward/reverse path. I wanna say, it's not very relevant for `/route`, so I personally wouldn't advocate to support what you'd like to see (even though it's totally doable).

I also say that, because you can do the filtering client-side, it's not necessary for Valhalla to filter those edges geometrically. You have the polygon and the response is giving you the road geometry. Hope that helps.

**Comment by OliverHering** [comment]
Thank you for your quick response.
But if I send the upper request to the route service, maybe with an additional parameter "search_cutoff"(?), like this:

`{
    "locations": [
        {
            "lat": 52.362323,
            "lon": 9.707735
        },
        {
            "lat": 52.354487,
            "lon": 9.73085,
            "search_cutoff": 200,
            "radius": 200
        }
    ],
    "costing": "bicycle",
    "exclude_polygons": [
        [
            [
                9.726981,
                52.3613
            ],
            [
                9.729886,
                52.362592
            ],
            [
                9.729927,
                52.362617
            ],
            [
                9.73777,
                52.361528
            ],
            [
                9.740011,
                52.356256
            ],
            [
                9.738185,
                52.352784
            ],
            [
                9.729512,
                52.352581
            ],
            [
                9.726981,
                52.3613
            ]
        ]
    ],
    "units": "km",
    "id": "123"
}`

Why do I not receive a response ("No path could be found for input")? The destination is within the exclude_polygon, but the radius should be high enough to find the nearest location outside this polygon? (If I understood it correctly, the locate function should provide sufficient ways in the first third of the route request.)

**Reply by nilsnolde** [reply]
It's hard to confirm for me what the request looks like.. And I don't have the time to look at it right now. Best you do some screenshots which show the locations & polygons. If I do it on https://valhalla.openstreetmap.de/ and draw a polygon around one of the locations, it's not working.

**Reply by OliverHering** [reply]
![valhalla_destination](https://user-images.githubusercontent.com/12051796/231994858-393c7899-1493-41b2-8d19-119b64afd421.png)
red : exclude_polygon
blue : 200m radius 
green : destination
